### ISSUE TITLE
Querying processes table the start time field is not always returning the same value for the same process pid and process instance

### ISSUE THREAD

**Issue by tmcg917** [osquery/osquery#6911]
# Bug report

### What operating system and version are you using?

ubuntu 18.04

### What version of osquery are you using?

4.5 and 4.6

### What steps did you take to reproduce the issue?
Ran the following query, filtering on a long running process, may have to test a few processes to see this issue.
Run the query several times, to see the start_time change by 1 second and then back again.

select pid, start_time from processes where pid = <pid for long running process>

### What did you expect to see?

The start_time epoch value to be the same, on every query run.

### What did you see instead?
Example output from osqueryi  after running the query several times, the start_time value is not consistent for the same process.

osquery> select pid, start_time from processes where pid = 19297;
-----------------+
pid	start_time
-----------------+
19297	1611187600
-----------------+
osquery> select pid, start_time from processes where pid = 19297;
-----------------+
pid	start_time
-----------------+
19297	1611187601
-----------------+
osquery> select pid, start_time from processes where pid = 19297;
-----------------+
pid	start_time
-----------------+
19297	1611187601
-----------------+
osquery> select pid, start_time from processes where pid = 19297;
-----------------+
pid	start_time
-----------------+
19297	1611187601
-----------------+
osquery> select pid, start_time from processes where pid = 19297;
-----------------+
pid	start_time
-----------------+
19297	1611187600
-----------------+
osquery>

**Comment by Smjert** [comment]
The problem is that we calculate the start time as `(now - uptime) + (proc.start_time / clock_ticks_per_second)`, and specifically `now` and `uptime` aren't taken atomically, so they are two different calls (one to `std::time()`, the other to `sysinfo()`) that sometimes can happen on a different second, and so they differ by one second.

Calculating `(now - uptime)` only once and storing it in a static variable should fix the issue, since the boot time won't change without the osquery process being started again.

**Comment by directionless** [comment]
> Calculating `(now - uptime)` only once and storing it in a static variable should fix the issue, since the boot time won't change without the osquery process being started again.

I remember something about uptime being inconsistent with how laptop sleep. Is that going to be a problem if this is a static assignment?

**Comment by Smjert** [comment]
> I remember something about uptime being inconsistent with how laptop sleep. Is that going to be a problem if this is a static assignment?

I don't think there are problems with that, I've also tested with suspend and hibernate and the uptime is correct.

Thinking about the issue more I'm realizing that it's actually more complex than initially thought and an unsolvable issue (if we want the start timestamp to never change). If we are going to save the boot timestamp once, there's no guarantee that the time was correct, because the clock could've been incorrect when read and then ntp could update the clock slightly.
Therefore one has to recalculate it everytime.

Meanwhile, I've noticed that `/proc/stat` contains the field `btime` which contains the boot time as a unix timestamp.
This might seem the solution, but it still somewhat has the original issue: if the time is changed between the acquisition of the `btime` and the acquisition of the process start time (seconds relative to the boot time), then we will have again an incorrect start timestamp, which will change at the next query.

I believe the only value that can stay fixed and can't change is what `/proc/<pid>/stat` reports as the start time of the process, which we currently use. That value represents the seconds passed since the boot time. That is retrieved once by the kernel, calculated as `now - boot time` and there's no ntp time change that can happen in between, so it's always correct.
But then, we have Windows which doesn't give you the seconds since the boot time (or any reference timestamp), but it gives the start timestamp, which can therefore change again due to ntp.

So in conclusion, I don't think it's possible to have a start time that never changes, but we can stabilize its value better and avoid frequent changes by reading the `btime` from proc, because then, only a ntp correction happening between the read of `btime` and the process start time can cause `start_time` to change.

**Comment by directionless** [comment]
It seems like in the face of NTP, it's very hard to reliably convert this to an absolute timestamp. So I think we should accept that there might be slight variation from NTP, rounding, etc. If we wanted higher precision, we could report the start_time field as the stat does -- as an elapsed interval. But I think we should do nothing and leave this as documentation

**Comment by tmcg917** [comment]
Some interesting comments, however the problem is that the start_time value provides a way to correctly map processes with the process_events table..  For example if child processes are started and shutdown within the same second as parent process started in.  The child processes may not be picked up in a query which combines processes and process_events tables.  Also we combine a pid and start_time to create a unique id across all tables, which is passed from one query into another.  

If the start_time changes, the information obtained from various queries cannot be trusted to be complete, therefore this issue needs a solution.

**Comment by theopolis** [comment]
One idea is to use the static calculation approach, yet allow updating from NTP and account for a small period of jitter.

For example:
- Calculate the boot time and store it statically.
- Continue to inspect the time on each time retrieval and if there is a delta beyond the accounted jitter (e.g., 2s) then assume NTP or some other factor has updated the time and re-store this new time.

This will make most calls to `start_time` stable, while still allowing NTP to work as intended.

**Comment by theopolis** [comment]
> Also we combine a pid and start_time to create a unique id across all tables

This sounds like a good approach but a more accurate one might include using `pid` and `tid` (thread ID) as there is more entropy in the thread IDs. However I am not sure if they are an OS agnostic concept and I would need to check the various tables to see if thr

**Comment by theopolis** [comment]
(sorry, I mis-clicked, I did not intend to close the issue)

**Comment by mathrock** [comment]
This back and forth "off by one" second issue with the `start_time` causes tons of unnecessary events for differential queries.  Every time the `start_time` jumps ahead one second it triggers a new diff query event (on the next scheduled query).  Same for when the `start_time` jumps back.  This causes constant unnecessary events for long running processes that aren't changing.

**Comment by mathrock** [comment]
Could a similar solution to fixing this on MacOS work for fixing this on Linux? #3853

**Comment by directionless** [comment]
Looking at #3892and dropping some notes here...

It creates a darwin-only `getUptimeInUSec` function. Which, for reasons unclear to me, ignores the reported boot usec, and calculates it based on gettimeofday

That said... It seems possible to create a cross platform `getUptimeInUSec`. Stack overflow has a reasonable looking [answer](https://stackoverflow.com/questions/30095439/how-do-i-get-system-up-time-in-milliseconds-in-c)

**Comment by tmcg917** [comment]
Any update for a fix for this issue.

**Comment by JayantNayak** [comment]
We are also seeing the same issue on macOS as well for Osqquery.
